"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeFn = void 0;
const errors_1 = require("./errors");
const serialize_error_1 = require("./serialize-error");
const executeFn = async (fn, args, authenticate, authContext) => {
    const start = Date.now();
    try {
        if (authenticate) {
            if (!authContext) {
                throw new errors_1.InferableError(errors_1.InferableError.JOB_AUTHCONTEXT_INVALID);
            }
            await authenticate(authContext, args[0]);
        }
        const result = await fn(...args);
        return {
            content: result,
            type: "resolution",
            functionExecutionTime: Date.now() - start,
        };
    }
    catch (e) {
        const functionExecutionTime = Date.now() - start;
        if (e instanceof Error) {
            return {
                content: (0, serialize_error_1.serializeError)(e),
                type: "rejection",
                functionExecutionTime,
            };
        }
        else if (typeof e === "string") {
            return {
                content: (0, serialize_error_1.serializeError)(new Error(e)),
                type: "rejection",
                functionExecutionTime,
            };
        }
        else {
            return {
                content: new Error("Inferable encountered an unexpected error type. Make sure you are throwing an Error object."),
                type: "rejection",
                functionExecutionTime,
            };
        }
    }
};
exports.executeFn = executeFn;
//# sourceMappingURL=execute-fn.js.map